### ISSUE TITLE
Training Time Discrepancy and Low GPU Utilization

### ISSUE THREAD

**Issue by VegaLighttt** [Aerial-Manipulation-Lab/MARL_mav_ext#25]
Hi, 
I have two questions I’d like to ask regarding the training time discrepancy and low GPU utilization.

**1. Training Time Discrepancy** Your paper mentions that training took **17 hours** on a single RTX 3090. However, following the instructions in `README.md`, I am experiencing significantly longer training times on these platforms:
- **100h** on an AMD EPYC 9T24 + Ada 5880.
- **70h** on a Xeon 8352V + RTX 4090.

Could you confirm if the 17h result in the paper was achieved using the exact same `num_envs` and `task` configurations as listed in the README?

**2. Low GPU Utilization & Optimization** I noticed that GPU utilization is very low when training `marl_hover`. The current structural design seems not suitable for GPU parallelism, so I wanted to ask if you have plans to optimize this to reduce CPU-GPU interaction overhead in the future?

It appears the original controller is derived from single-agent. When extended to the multi-agent setting, it seems the controller is called for each UAV individually, executing serially within a `for` loop. I ran a profiler and found that `env_step` accounts for 57.81% of the CPU time. Most of the environment simulation logic seems to be executed serially on the CPU.

I attempted several optimizations—such as avoiding `torch.tensor` creation at every `env_step`, minimizing tensor copying between CPU and GPU, and batching the controller. However, the training time only improved by about 10-15%.

**Comment by jackzeng-robotics** [comment]
Hi @VegaLighttt,

1. Those seem like very long training times, yes I confirm that 17h was with our setup on a RTX 3090, with 4096 parallel environments. You should quickly see the task rewards and average episode length go up. The task specific objectives are usually optimized first, before taking a longer time to learn the action smoothness rewards. Are you sure you are using the right amount of envs and the right hyperparameter setup? Even with 5 drones it should not take longer than ~24 hours.

2. I have also used a profiler, but it is true that the code is not fully optimized for efficiency. We also ran experiments with a variable amount of drones, so the could is partly built for modularity to test different setups. The for loop is one of the things that could be optimized yes. This does not mean however that the code is ran on the CPU, as the tensors are still handled on the GPU. 

With the long training times in consideration, perhaps there are some settings in your local isaac sim/lab installation that prevent the task from running smoothly? Our CPU was Intel core i9, but I doubt that that was the enabler, since the training would also run within a reasonable amount of time on my local laptop with an RTX 2080 (with fewers envs).

**Comment by VegaLighttt** [comment]
I really appreciate you taking the time to answer my questions.
I have configured the environment following the `README.md` and am currently on the latest git commit `f9d5cec` (HEAD -> main). I used the training command mentioned in the documentation:
```bash
python3 scripts/skrl/train.py --task=Isaac-flycrane-payload-decentralized-hovering-v0 --headless --num_envs=4096 --seed=-1 --algorithm="MAPPO"
```
Although the project environment is Isaac Sim 4.5 + Isaac Lab 2.0.2, I have attempted to reproduce the results on different servers with various configurations:
- Isaac Sim 4.5 / Lab 2.1.0
- Isaac Sim 5.0 / Lab 2.2.0 (Incompatible)
- Isaac Sim 4.5 / Lab 2.0.0
Additionally, I set up the same environment(Sim4.5/Lab 2.0.2) with i9 CPU + RTX 4090 and ran the same command. While the rewards do increase quickly during training, the total training time has not improved to the expected ~17 hours. The fastest time I achieved (on the i9 + 4090) was still nearly 40 hours.

I manage Isaac Sim/Lab using Conda. The only deviation from the standard process is that I encountered a `toml` error when installing `MARL_mav_carry_ext`. Then, I used `pip install -e . --no-build-isolation` However, I think that is not the cause of such a time gap.

Paper Figure 11 indicates a training range of 0-1000M environment steps. However, the `skrl_mappo_cfg.yaml` (line 98) specifies `timesteps: 1500000`. Assuming the goal is 1000M environment steps, this implies the parallel environments are approximately 667. With this configuration, yes, the training duration is estimated to be 17-18 hours(i9+4090).

I'm confused by the performance difference. Based on my current speed on an i9 + 4090, I have to drop the parallel environments down to 667 to finish in 17-18 hours. Since you managed to run 4096 environments in that same timeframe, I'm wondering if I missed any specific optimizations you used to get that speed?

**Comment by jackzeng-robotics** [comment]
Hi @VegaLighttt, I do not have my logs at hand anymore but perhaps you could check if the bottleneck is in the simulation time (data collection), or the learning time. In my fork of SKRL I should have added those metrics. Again, I do not have the numbers at hand but that could be possible cause. Also, applying advantage filtering as stated in the paper sped up training by a lot (this should also be enabled already in my fork).
